hadoop-hdfs-project/hadoop-hdfs-rbf/src/test/java/org/apache/hadoop/hdfs/server/federation/router/TestNoNamenodesAvailableLongTime.java

```diff
@@ -0,0 +1,432 @@
+  private boolean routerCacheNoActiveNamenode(
+      RouterContext context, String nsId, boolean useObserver) throws IOException {
+    List<? extends FederationNamenodeContext> namenodes
+        = context.getRouter().getNamenodeResolver().getNamenodesForNameserviceId(nsId, useObserver);
+    for (FederationNamenodeContext namenode : namenodes) {
+      if (namenode.getState() == FederationNamenodeServiceState.ACTIVE){
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * All routers in the cluster force loadcache.
+   */
+  private void allRoutersLoadCache() {
+    for (MiniRouterDFSCluster.RouterContext context : cluster.getRouters()) {
+      // Update service cache
+      context.getRouter().getStateStore().refreshCaches(true);
+    }
+  }
+
+  /**
+   * Set the second non-observer state namenode in the router cache to active.
+   */
+  private void setSecondNonObserverNamenodeInTheRouterCacheActive(
+      int numberOfObserver, boolean useObserver) throws IOException {
+    List<? extends FederationNamenodeContext> ns0 = routerContext.getRouter()
+        .getNamenodeResolver()
+        .getNamenodesForNameserviceId("ns0", useObserver);
+
+    String nsId = ns0.get(numberOfObserver+1).getNamenodeId();
+    cluster.switchToActive("ns0", nsId);
+    assertEquals(ACTIVE.ordinal(),
+        cluster.getNamenode("ns0", nsId).getNamenode().getNameNodeState());
+
+  }
+
+  /**
+   * All routers in the cluster force heartbeat.
+   */
+  private void allRoutersHeartbeat() throws IOException {
+    for (RouterContext context : cluster.getRouters()) {
+      // Manually trigger the heartbeat, but the router does not manually load the cache
+      Collection<NamenodeHeartbeatService> heartbeatServices = context
+          .getRouter().getNamenodeHeartbeatServices();
+      for (NamenodeHeartbeatService service : heartbeatServices) {
+        service.periodicInvoke();
+      }
+    }
+  }
+
+  /**
+   * Transition the active namenode in the cluster to standby.
+   */
+  private void transitionActiveToStandby() {
+    if (cluster.isHighAvailability()) {
+      for (String ns : cluster.getNameservices()) {
+        List<MiniRouterDFSCluster.NamenodeContext>  nnList = cluster.getNamenodes(ns);
+        for (MiniRouterDFSCluster.NamenodeContext namenodeContext : nnList) {
+          if (namenodeContext.getNamenode().isActiveState()) {
+            cluster.switchToStandby(ns, namenodeContext.getNamenodeId());
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Shutdown observer namenode in the cluster.
+   *
+   * @param num The number of shutdown observer.
+   */
+  private void stopObserver(int num) {
+    int nnIndex;
+    for (nnIndex = 0; nnIndex < cluster.getNamenodes().size(); nnIndex++) {
+      NameNode nameNode = cluster.getCluster().getNameNode(nnIndex);
+      if (nameNode != null && nameNode.isObserverState()) {
+        cluster.getCluster().shutdownNameNode(nnIndex);
+        num--;
+        if (num == 0) {
```

Comment:
Can you just do it in the for loop?
```
int numNns = cluster.getNamenodes().size();
for (nnIndex = 0; nnIndex < numNns && num > 0; nnIndex++) {
```
It is also a little weird to modify an input parameter.

Reply:
Thanks for your advice, done